src/metaopt/io/parsing.py

```diff
@@ -0,0 +1,257 @@
+    database=ENV_VARS_DB
+    )
+
+################################################################################
+#                           Input Parsing Functions                            #
+################################################################################
+
+
+def mopt_args(description):
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description=textwrap.dedent(description))
+
+    parser.add_argument(
+        '-V', '--version',
+        action='version', version='metaopt ' + metaopt.__version__)
+
+    parser.add_argument(
+        '-v', '--verbose',
+        action='count', default=0,
+        help="logging levels of information about the process (-v: INFO. -vv: DEBUG)")
+
+    moptgroup = parser.add_argument_group(
+        "MetaOpt arguments (optional)",
+        description="These arguments determine mopt's behaviour")
+
+    moptgroup.add_argument(
+        '-e', '--exp-name',
+        type=str, metavar='stringID',
+        help="experiment's unique name; "
+             "use an existing name to resume an experiment "
+             "(default: %s)" % DEF_CMD_EXP_NAME[1])
+
+    moptgroup.add_argument(
+        '--max-trials', type=int, metavar='#',
+        help="number of jobs/trials to be completed "
+             "(default: %s)" % DEF_CMD_MAX_TRIALS[1])
+
+    moptgroup.add_argument(
+        "--pool-size", type=int, metavar='#',
+        help="number of concurrent workers to evaluate candidate samples "
+             "(default: %s)" % DEF_CMD_POOL_SIZE[1])
+
+    moptgroup.add_argument(
+        '-mc', '--moptconfig',
+        type=argparse.FileType('r'), metavar='path-to-config',
+        help="user provided mopt configuration file")
+
+    usergroup = parser.add_argument_group(
+        "User script related arguments",
+        description="These arguments determine user's script behaviour "
+                    "and they can serve as metaopt's parameter declaration.")
+
+    usergroup.add_argument(
+        '-c', '--userconfig',
+        type=str, metavar='path-to-config',
+        help="your script's configuration file (yaml, json, ini, ...anything)")
+
+    usergroup.add_argument(
+        'userscript', type=str, metavar='path-to-script',
+        help="your experiment's script")
+
+    usergroup.add_argument(
+        'userargs', nargs=argparse.REMAINDER, metavar='...',
+        help="command line arguments to your script (if any)")
+
+    args = vars(parser.parse_args())  # convert to dict
+    # Explicitly add metaopt's version as experiment's metadata
+    args['version'] = metaopt.__version__
+    moptfile = args.pop('moptconfig')
+    config = dict()
+    if moptfile:
+        config = yaml.safe_load(moptfile)
+
+    if args['verbose'] == 1:
+        logging.basicConfig(level=logging.INFO)
+    elif args['verbose'] == 2:
+        logging.basicConfig(level=logging.DEBUG)
+
+    return args, config
+
+
+def default_options(user, starttime):
```

Comment:
> To make things clear,...MetaOpt configuration optional.

That's my intention too. I think it successfully does that.

> For this distinction, I would... script outputs, configurations, etc.

Yeah, I could not think of a better name 😛. Probably `resolveconfig` (even within `metaopt.io`) would be more appropriate I think. Do you agree?

Regarding config files, I was thinking something like this:

```yaml
database: {
    # some global definitions
}
resources: {
    # some global definitions
}
optimizer: {
    # `Optim`, `Dynamic` stuff here
}
# other common experiment metadata follow here (each one overrides the previous one in hierarchy)
```

I would not like parameters of [black-box] functions to be coupled to optimization procedures. So that the same procedure definition, i.e. cmd-provided moptconfig file + default files, can be re-used across different problems, i.e. userscript/[black box] function + hyperparameters/function domain. Although, in theory some optimization procedures are more appropriate for some function classes, I think we should act as if we were ignorant towards this and allow the user to re-use definitions according to their beliefs for "appropriate" correspondences.

I think that user provided defaults and config files are meant to represent his particular beliefs about best optimization procedures (`optimizer` key), his usual resource preferences (`resources` key - such as scheduler args for which queue he usually submits his jobs), and personal workflow (`database` key - mby he keeps a personal database).

I am pretty positive that we can be config-type agnostic towards hparam space definition, so that we let user keep whatever they used to use, and use their config file both as declaration and as template + their script's cmd args. Yes, I agree that any hparams declaration found will definitely be logged in experiment's db entry and thus there would be a dictionary representing that at some point.

Reply:
> Yeah, I could not think of a better name :stuck_out_tongue:. Probably resolveconfig (even within metaopt.io) would be more appropriate I think. Do you agree?

Yes, resolve_config would be even better. :wink: 

It think there is two concepts here which are currently entangled but should not dictate the design of each other. There is the concept of *hierarchical* or *distributed* configuration which enables global definitions of DB, resources and optimizer to favor reuse over many experiments. The second concept is the support for any type of text-based configuration file for the hyper-parameter space. 

### *Hierarchical* or *distributed* configuration

I do also believe having support for flexible configuration is important. However, I also believe default options for Optimizer will address most of the use cases. Furthermore, from my experience, people tend to keep experiment definitions centralized and prefer to copy-paste them and edit rather than refactoring it. By centralizing, I mean having all the configuration in a single file rather than many layers of global to exp-specific definitions. We could argue like for the git-commit message/exp-name what is in their best interest. In this case, the main argument for keeping config centralized is for having something easy to reproduce and also a simple reference for later. DB and ressources config and not crucial for reproducibility, for optimizer is. I find this workflow well adapted to research and I don't think we should discourage it.

I should also note that some argument of the optimizer are very problem specific. There is for exemple the type of optimizer (minimize, maximize), the type of task to optimizer (noiseless, gaussian, etc), some constraints on the hyper-parameters like in [spearmint's tasks](https://github.com/HIPS/Spearmint/blob/master/examples/constrained/config.json#L19-L32). All of them can be optional, with default values covering a large part of the use cases, but when users need to set them, that is unlikely to be very reusable. 

To be clear, I would not recommend to drop support for hierarchical/distributed configuration, but it is my belief that it should not be enforced on users. They should be able to use a single config file if they prefer without the complexity hassle of hierarchical/distributed config. It should be transparent. I believe your code presently support this beside hyper-parameter space, and this brings us to next point. 

### Support for any text-based hyper-parameter space definition file type

This feature is very nice. I can think of python scripts for instance, something I've seen often in others pipelines, that could be templates for hyper-parameters settings. Very neat. This feature however should not force us to keep hyper-parameter space definitions outside of the configuration file. Again, I believe a very common use case will be to simply use the experiment configuration file with the optimizer definition beside the hyper-parameter space definition. This would not cause any problem for agnostic support for text-based configs. We can simply update the `hp_space` dictionary from the experiment configuration file with a h-p space configuration file given in command-line. 

All in all, those two concepts are nice and I agree should be part of the project. They are however features that a minimal viable product can live without. In case of the hierarchical configuration, it seems you are close to having it built entirely already. In that case it might be worth finishing right now. For the agnostic h-p space parser I would leave it for later. You should open new issues for each features like this and give them a detailed description so that we do not lose the idea.